lib/datadog/appsec/contrib/graphql/gateway/multiplex.rb

```diff
@@ -28,41 +28,74 @@ def queries
+            #     title
+            #     comments @include(if: $withComments) {
+            #       author { name }
+            #       content
+            #     }
+            #   }
+            # }
+            #
+            # The result would be:
+            # {"post"=>[{"slug"=>"my-first-post"}], "comments"=>[{"include"=>{"if"=>true}}]}
+            #
+            # Note that the `comments` "include" directive is included in the arguments list
+            def build_arguments_hash
+              queries.each_with_object({}) do |query, args_hash|
+                next unless query.selected_operation
+
+                arguments_from_selections(query.selected_operation.selections, query.variables, args_hash)
+              end
+            end
+
+            def arguments_from_selections(selections, query_variables, args_hash)
+              selections.each do |selection|
+                next unless selection.is_a?(::GraphQL::Language::Nodes::Field)
+
+                selection_name = selection.alias || selection.name
+
+                if selection.arguments.any? || selection.directives.any?
+                  args_hash[selection_name] ||= []
+                  args_hash[selection_name] <<
+                    arguments_hash(selection.arguments, query_variables).merge(
+                      arguments_from_directives(selection.directives, query_variables)
+                    )
                 end
-                next if resolver_args.empty? && resolver_dirs.empty?
 
-                args_resolver = (args[query.operation_name || "query#{index + 1}"] ||= [])
-                # We don't want to add empty hashes so we check again if the arguments and directives are empty
-                args_resolver << resolver_args unless resolver_args.empty?
-                args_resolver << resolver_dirs unless resolver_dirs.empty?
+                arguments_from_selections(selection.selections, query_variables, args_hash)
               end
-              args
             end
 
-            # Set the resolver hash (resolver_args and resolver_dirs) with the arguments and provided variables
-            def set_hash_with_variables(resolver_hash, arguments, provided_variables)
-              arguments.each do |arg|
-                resolver_hash[arg.name] =
-                  if arg.value.is_a?(::GraphQL::Language::Nodes::VariableIdentifier)
-                    provided_variables[arg.value.name]
-                  else
-                    arg.value
-                  end
+            def arguments_from_directives(directives, query_variables)
+              directives.map do |directive|
+                next unless directive.is_a?(::GraphQL::Language::Nodes::Directive)
+
+                [directive.name, arguments_hash(directive.arguments, query_variables)]
+              end.to_h
+            end
+
+            def arguments_hash(arguments, query_variables)
+              arguments.map do |argument|
```

Comment:
if we use more stupid solution

```ruby
arguments.each_with_object({})  do |argument, hash|
  hash[argument.name] = argument_value(argument, query_variables)
end
```

it will save N+1 array allocations in memory, which on a scale [could be a lot](https://mensfeld.pl/2024/08/under-the-hood-enhancing-karafkas-cpu-and-memory-efficiency)

Reply:
Fixed here: [d51cc17](https://github.com/DataDog/dd-trace-rb/pull/3887/commits/d51cc170da1a913520719904697ed343f43d51ad)